Backpack/src/main/java/net/skyscanner/backpack/contentbubble/BpkContentBubble.kt

```diff
@@ -0,0 +1,224 @@
+        _round = it.getBoolean(R.styleable.BpkContentBubble_contentBubbleRound, round)
+
+        _pointerPosition = it.getInt(R.styleable.BpkContentBubble_contentBubblePointerPosition, pointerPosition.id)
+          .let(::mapXmlToPointerPosition) ?: pointerPosition
+      }
+
+    background = null
+    setWillNotDraw(false)
+    setPadding(0, 0, 0, 0)
+    // PorterDuffXfermode doesn't work in some devices when hardware acceleration is on
+    setLayerType(LAYER_TYPE_SOFTWARE, null)
+
+    pointerDrawable.apply {
+      setBounds(0, 0, intrinsicWidth, intrinsicHeight)
+    }
+
+    cornerRadiusDrawable.apply {
+      setBounds(0, 0, intrinsicWidth, intrinsicHeight)
+    }
+
+    setUpMasks()
+  }
+
+  override fun draw(canvas: Canvas) {
+    val count = canvas.saveCount
+
+    val width = width.toFloat()
+    val height = height.toFloat()
+
+    val pointerHalfWidth = pointerDrawable.intrinsicWidth / 2
+    val pointerOffset = if (layoutDirection == View.LAYOUT_DIRECTION_RTL) {
+      1 - pointerPosition.offset
+    } else {
+      pointerPosition.offset
+    }
+
+    val yStart = height - pointerDrawable.intrinsicHeight
+    val pointerXStart = width * pointerOffset - pointerHalfWidth
+    val pointerXEnd = width * pointerOffset + pointerHalfWidth
+
+    clipRect.set(0f, yStart, pointerXStart, height)
+    canvas.clipOutRectCompat(clipRect)
+
+    clipRect.set(width, yStart, pointerXEnd, height)
+    canvas.clipOutRectCompat(clipRect)
+
+    super.draw(canvas)
+    canvas.restoreToCount(count)
+
+    paint.xfermode = porterDuffXfermode
+    canvas.drawBitmap(mask, pointerXStart, yStart, paint)
+
+    if (round) {
+      drawRadiusMask(canvas)
+    }
+
+    paint.xfermode = null
+  }
+
+  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+    val firstChild = getChildAt(0)
+    if (firstChild == null || !fitContent || firstChild.layoutParams.height >= 0) {
+      super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+    } else {
+      getChildAt(0)?.let {
+        // When fit content is true we add the size of the pointer to the view's height, in order
+        // for the children to take up that size we need to change `WRAP_CONTENT` to `MATCH_PARENT`
+        if (it.layoutParams.height == ViewGroup.LayoutParams.WRAP_CONTENT) {
+          it.layoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
+        }
+        it.measure(widthMeasureSpec, heightMeasureSpec)
+      }
+
+      super.onMeasure(
+        widthMeasureSpec,
+        MeasureSpec.makeMeasureSpec(
+          getChildAt(0).measuredHeight + pointerDrawable.intrinsicHeight,
+          MeasureSpec.EXACTLY
+        ))
+    }
+  }
+
+  private fun Canvas.clipOutRectCompat(rect: RectF) {
+    if (Build.VERSION.SDK_INT >= 26) {
+      this.clipOutRect(rect)
+    } else {
+      @Suppress("DEPRECATION")
+      this.clipRect(rect, Region.Op.DIFFERENCE)
+    }
+  }
+
+  private fun setUpMasks() {
```

Comment:
This logic might be moved into static lazy init getters, e.g. getMask(Context).

Reply:
And how would you declare this?  Making a function that receives an argument lazy requires us to do that manually, and also make sure we do it per context. At this point we will be basically reimplementing the internal cache for no reason.

Reply:
I did not change this is will be a premature optimisation that may cause other problems for no benefit. This view is intended to be used once in every page so memory wise this will not make a difference and we can rely on the Android internal cache for the drawables